### ISSUE TITLE
Cant build zstd-dll with 1.5.4

### ISSUE THREAD

**Issue by nolange** [facebook/zstd#3492]
**Describe the bug**

**To Reproduce**

This is on debian testing x86

``` bash
make -j9 PREFIX=/usr ZSTD_LEGACY_SUPPORT=0 HAVE_ZLIB=0 HAVE_LZMA=0 HAVE_LZ4=0 MOREFLAGS="-Os" HAVE_THREAD=1 -C lib libzstd.pc-mt-release lib-mt-release
make -j9 PREFIX=/usr ZSTD_LEGACY_SUPPORT=0 HAVE_ZLIB=0 HAVE_LZMA=0 HAVE_LZ4=0 MOREFLAGS="-Os" HAVE_THREAD=1 -C programs zstd-dll
```

**Expected behavior**

zstd library and program being build (like it worked with 1.5.2)

**Actutal behavior**

```
LINK obj/conf_2f7cbea6a959ba80ad7e06b8fde79417/zstd
cc -DXXH_NAMESPACE=ZSTD_ -DDEBUGLEVEL=0 -DZSTD_LEGACY_SUPPORT=0 -DZSTD_MULTITHREAD    -DZSTD_LEGACY_SUPPORT=0 -O3 -Wall -Wextra -Wcast-qual -Wcast-align -Wshadow -Wstrict-aliasing=1 -Wswitch-enum -Wdeclaration-after-statement -Wstrict-prototypes -Wundef -Wpointer-arith -Wvla -Wformat=2 -Winit-self -Wfloat-equal -Wwrite-strings -Wredundant-decls -Wmissing-prototypes -Wc++-compat -Os -Wa,--noexecstack -Wall -Wextra -Wcast-qual -Wcast-align -Wshadow -Wstrict-aliasing=1 -Wswitch-enum -Wdeclaration-after-statement -Wstrict-prototypes -Wundef -Wpointer-arith -Wvla -Wformat=2 -Winit-self -Wfloat-equal -Wwrite-strings -Wredundant-decls -Wmissing-prototypes -Wc++-compat -Os -O3 -Wall -Wextra -Wcast-qual -Wcast-align -Wshadow -Wstrict-aliasing=1 -Wswitch-enum -Wdeclaration-after-statement -Wstrict-prototypes -Wundef -Wpointer-arith -Wvla -Wformat=2 -Winit-self -Wfloat-equal -Wwrite-strings -Wredundant-decls -Wmissing-prototypes -Wc++-compat -Os -Wa,--noexecstack -Os -z noexecstack -L../lib -pthread  obj/conf_2f7cbea6a959ba80ad7e06b8fde79417/xxhash.o obj/conf_2f7cbea6a959ba80ad7e06b8fde79417/benchfn.o obj/conf_2f7cbea6a959ba80ad7e06b8fde79417/benchzstd.o obj/conf_2f7cbea6a959ba80ad7e06b8fde79417/datagen.o obj/conf_2f7cbea6a959ba80ad7e06b8fde79417/dibio.o obj/conf_2f7cbea6a959ba80ad7e06b8fde79417/fileio.o obj/conf_2f7cbea6a959ba80ad7e06b8fde79417/fileio_asyncio.o obj/conf_2f7cbea6a959ba80ad7e06b8fde79417/timefn.o obj/conf_2f7cbea6a959ba80ad7e06b8fde79417/util.o obj/conf_2f7cbea6a959ba80ad7e06b8fde79417/zstdcli.o obj/conf_2f7cbea6a959ba80ad7e06b8fde79417/zstdcli_trace.o -lzstd    -o obj/conf_2f7cbea6a959ba80ad7e06b8fde79417/zstd
/usr/bin/ld: obj/conf_2f7cbea6a959ba80ad7e06b8fde79417/fileio_asyncio.o: in function `AIO_IOPool_destroy':
fileio_asyncio.c:(.text+0x4ed): undefined reference to `POOL_free'
/usr/bin/ld: obj/conf_2f7cbea6a959ba80ad7e06b8fde79417/fileio_asyncio.o: in function `AIO_IOPool_init':
fileio_asyncio.c:(.text+0x63f): undefined reference to `POOL_create'
/usr/bin/ld: obj/conf_2f7cbea6a959ba80ad7e06b8fde79417/fileio_asyncio.o: in function `AIO_IOPool_enqueueJob':
fileio_asyncio.c:(.text+0x13d): undefined reference to `POOL_add'
/usr/bin/ld: obj/conf_2f7cbea6a959ba80ad7e06b8fde79417/fileio_asyncio.o: in function `AIO_IOPool_join':
fileio_asyncio.c:(.text+0x3ed): undefined reference to `POOL_joinJobs'
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:162: obj/conf_2f7cbea6a959ba80ad7e06b8fde79417/zstd] Error 1
make: *** [Makefile:150: zstd] Error 2
make: Leaving directory '/tmp/zstd-1.5.4/programs'
```

**Desktop (please complete the following information):**
 - OS: Debian Linux
 - Version:  Bookworm x86_64
 - Compiler: gcc (Debian 12.2.0-14) 12.2.0 
 - Flags: -Os
 - Build system: Makefile

**Additional context**

**Comment by nolange** [comment]
This would be a possible fix:

``` diff
--- programs/Makefile.org	2023-02-10 13:28:17.294882218 +0100
+++ programs/Makefile	2023-02-10 13:28:22.134841461 +0100
@@ -224,7 +224,7 @@
 ## zstd-dll: zstd executable linked to dynamic library libzstd (must have same version)
 .PHONY: zstd-dll
 zstd-dll : LDFLAGS+= -L$(LIBZSTD)
-zstd-dll : LDLIBS += -lzstd
+zstd-dll : LDLIBS += -lzstd $(LIBZSTD)/libzstd.a
 zstd-dll : ZSTDLIB_LOCAL_SRC = xxhash.c
 zstd-dll : zstd
```

**Comment by yoniko** [comment]
@nolange - thank you for reporting, we will be looking into it.

**Comment by yoniko** [comment]
Fixed in #3496.
@nolange - your solution actually links the static library into the zstd binary, removing the dynamic dependency altogether making it the same as static Zstd build. I'd suggest using the patch from #3498 or using current dev if you need to build with dynamic library.

**Comment by nolange** [comment]
Na, link order matters. You pull in the DSO first, then the remaining stuff is used from the static library.

But I know it wasn't a thorough solution, thanks
